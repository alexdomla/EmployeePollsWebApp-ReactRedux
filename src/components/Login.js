// Import necessary dependencies
import React from "react";
import { connect } from "react-redux";
import { useState } from "react";
import { handleLogin } from "../actions/authedUser";
import { useNavigate } from "react-router-dom";

// Define the Login component
const Login = ({ dispatch, isLoggedIn }) => {
  // Define the navigate function from react-router-dom
  const navigate = useNavigate();

  // Initialize state variables using the useState hook
  const [username, setUsername] = useState("sarahedo");
  const [password, setPassword] = useState("password123");
  const [error, setError] = useState(false);
  const [success, setSuccess] = useState(false);

  // Handle changes to the username input field
  const handleUsername = (e) => {
    setUsername(e.target.value);
  };

  // Handle changes to the password input field
  const handlePassword = (e) => {
    setPassword(e.target.value);
  };

  // Handle form submission
  const handleSubmit = (e) => {
    e.preventDefault();
    if (!username || !password) {
      // If username or password is not entered, show error message
      setSuccess(false);
      setError(true);
      return;
    }

    // Dispatch action to log in user with entered credentials
    dispatch(handleLogin(username, password));
    navigate("/");

    // Reset username and password state variables and remove error message
    setUsername("");
    setPassword("");
    setError(false);
  };

  // Render the Login component
  return (
    <div className="loginPage">
      <h1>Login</h1>
      <div className="login__img">
        <img src="./images/avatars.png" alt="" />
      </div>
      {error && (
        <p data-testid="error-header">
          Please enter a correct username and a password.
        </p>
      )}
      <form onSubmit={handleSubmit}>
        <label htmlFor="">User</label>
        <input
          type="text"
          id="fname"
          name="fname"
          onChange={handleUsername}
          value={username}
          data-testid="name-input"
        ></input>
        <label htmlFor="">Password</label>
        <input
          type="text"
          id="fname"
          name="fname"
          onChange={handlePassword}
          value={password}
          data-testid="password-input"
        ></input>
        <button type="submit" data-testid="submit-button">
          Submit
        </button>
      </form>
    </div>
  );
};

// Map the authedUser state to props and export the component
const mapStateToProps = ({ authedUser }) => ({
  isLoggedIn: !!authedUser,
});

export default connect(mapStateToProps)(Login);
