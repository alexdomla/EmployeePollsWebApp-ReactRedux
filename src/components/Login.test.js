import Login from "../components/Login";
import { store } from "../store";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { fireEvent, render } from "@testing-library/react";

describe("Login", () => {
  it("will verify that the login component is rendering correctly", () => {
    const component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it("will display an error if the name or the password is not provided.", () => {
    const component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    var inputName = component.getByTestId("name-input");
    fireEvent.change(inputName, { target: { value: "" } });
    var inputPassword = component.getByTestId("password-input");
    fireEvent.change(inputPassword, { target: { value: "" } });
    var submitButton = component.getByTestId("submit-button");
    fireEvent.click(submitButton);
    expect(component.getByTestId("error-header")).toBeInTheDocument();
  });

  it("will not display an error if the name and the password are provided.", () => {
    const component = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );
    var inputName = component.getByTestId("name-input");
    fireEvent.change(inputName, { target: { value: "sarahedo" } });
    var inputPassword = component.getByTestId("password-input");
    fireEvent.change(inputPassword, { target: { value: "password123" } });
    var submitButton = component.getByTestId("submit-button");
    fireEvent.click(submitButton);
    expect(component.queryByTestId("error-header")).not.toBeInTheDocument();
  });
});
