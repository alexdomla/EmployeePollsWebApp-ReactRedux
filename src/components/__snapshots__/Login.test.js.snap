// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will verify that the login component is rendering correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="loginPage"
      >
        <h1>
          Login
        </h1>
        <div
          class="login__img"
        >
          <img
            alt=""
            src="./images/avatars.png"
          />
        </div>
        <form>
          <label
            for=""
          >
            User
          </label>
          <input
            data-testid="name-input"
            id="fname"
            name="fname"
            type="text"
            value="sarahedo"
          />
          <label
            for=""
          >
            Password
          </label>
          <input
            data-testid="password-input"
            id="fname"
            name="fname"
            type="text"
            value="password123"
          />
          <button
            data-testid="submit-button"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="loginPage"
    >
      <h1>
        Login
      </h1>
      <div
        class="login__img"
      >
        <img
          alt=""
          src="./images/avatars.png"
        />
      </div>
      <form>
        <label
          for=""
        >
          User
        </label>
        <input
          data-testid="name-input"
          id="fname"
          name="fname"
          type="text"
          value="sarahedo"
        />
        <label
          for=""
        >
          Password
        </label>
        <input
          data-testid="password-input"
          id="fname"
          name="fname"
          type="text"
          value="password123"
        />
        <button
          data-testid="submit-button"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
